xla/service/gpu/ir_emitter_unnested.cc

```diff
@@ -2557,61 +2558,83 @@ static std::optional<GlobalDeviceId> DeviceConstraint(
-          static_cast<int>(stream_executor::MemoryType::kHost)) {
-    VLOG(3) << "Device to Host: host memory space "
-            << static_cast<int>(stream_executor::MemoryType::kHost);
-    auto thunk = std::make_unique<DeviceToHostCopyThunk>(
-        Thunk::ThunkInfo::WithProfileAnnotation(instr),
-        /*source_buffer=*/src_buffer,
-        /*destination_buffer=*/dst_buffer,
-        /*mem_size=*/ShapeUtil::ByteSizeOf(input_shape));
-    AddThunkToThunkSequence(std::move(thunk));
-    return absl::OkStatus();
-  }
-  if (shape.mutable_tuple_shapes(1)->has_layout() &&
+          host_memory_space) {
+    dir = D2H;
+  } else if (shape.mutable_tuple_shapes(1)->has_layout() &&
       shape.mutable_tuple_shapes(1)->mutable_layout()->memory_space() ==
-          static_cast<int>(stream_executor::MemoryType::kHost)) {
-    VLOG(3) << "Host to Device from the host memory space "
-            << static_cast<int>(stream_executor::MemoryType::kHost);
-    ;
-    auto thunk = std::make_unique<HostToDeviceCopyThunk>(
+          host_memory_space) {
+    dir = H2D;
+  }
+  if (dir != D2D) {
+    auto thunk = std::make_unique<DeviceHostCopyThunk>(
         Thunk::ThunkInfo::WithProfileAnnotation(instr),
         /*source_buffer=*/src_buffer,
         /*destination_buffer=*/dst_buffer,
-        /*mem_size=*/ShapeUtil::ByteSizeOf(input_shape));
+        /*mem_size=*/ShapeUtil::ByteSizeOf(input_shape),
+        /*async_events=*/copy_events_,
+        /*copy_start_instr=*/instr,
+        /*device_to_host=*/static_cast<bool>(dir));
     AddThunkToThunkSequence(std::move(thunk));
     return absl::OkStatus();
   }
+  return absl::InternalError(
+      "Copy-start doesn't have correct host memory space color");
+}
 
-  // Disabled the generation of memcpy D2D as only H2D and D2H are useful
-  // for memory offload now.
+// In the CopyDone thunk, the corresponding copy start instruction is passed
+// in order to finding the matching event and make sure it's completed.
+absl::Status IrEmitterUnnested::EmitCopyDoneThunk(const HloInstruction *instr) {
+  TF_ASSIGN_OR_RETURN(BufferAllocation::Slice dst_buffer,
+                      GetAllocationSliceForHlo(instr,
+                                               /*ShapeIndex=*/{}));
 
-  auto thunk = std::make_unique<DeviceToDeviceCopyThunk>(
-      Thunk::ThunkInfo::WithProfileAnnotation(instr),
-      /*source_buffer=*/src_buffer,
-      /*destination_buffer=*/dst_buffer,
-      /*mem_size=*/ShapeUtil::ByteSizeOf(input_shape));
-  AddThunkToThunkSequence(std::move(thunk));
+  const HloInstruction *src = instr->operand(0);
+  const Shape shape = src->shape();
 
-  return absl::OkStatus();
+  TF_ASSIGN_OR_RETURN(BufferAllocation::Slice src_buffer,
+                      GetAllocationSliceForHlo(src, {0}));
+  CHECK(shape.IsTuple());
+
+  if ((shape.tuple_shapes(0).has_layout() &&
+       shape.tuple_shapes(0).layout().memory_space() ==
+           static_cast<int>(stream_executor::MemoryType::kHost)) ||
```

Comment:
Maybe a helper HasHostMemorySpace(Shape) or similar to avoid redundancy?

Reply:
Yes, added the helper HasHostMemorySpace(Shape).